File: String/Python/capitaliseString.py
# Given a string, return whether or not it uses capitalization correctly. 
# A string correctly uses capitalization if all letters are capitalized, 
# no letters are capitalized, or only the first letter is capitalized.

# Ex: Given the following strings...

# "USA", return true
# "Calvin", return true
# "compUter", return false
# "coding", return true

def capitaliseString(inputstr):
    allCaps = False
    firstCap = True
    noneCap = False
    for i in range(len(inputstr)):
        if i == 0 and not inputstr[i].isupper():
            firstCap = False
            noneCap = True
        if i == 1 and inputstr[i].isupper():
            allCaps = True

        if allCaps and inputstr[i].islower():
            return False
        if noneCap and inputstr[i].isupper():
            return False
        if not i == 0 and not allCaps and inputstr[i].isupper():
            return False
    return True


def test():
    testCases = ["USA", "Calvin", "compUter", "coding", "CCd", "CdC", "CCdE", "eeC", "eCe", "eeCe"]
    expecteds = [True, True, False, True, False, False, False, False, False, False]
    for case, expected in zip(testCases, expecteds):
        actual = capitaliseString(case)
        assert actual == expected, f"Wrong answer. Case: {case}. Expected: {expected}. Actual: {actual}."
    print("All test cases passed.")


test()

File: String/JavaScript/capitaliseString.js
// Given a string, return whether or not it uses capitalization correctly.
// A string correctly uses capitalization if all letters are capitalized,
// no letters are capitalized, or only the first letter is capitalized.

// Ex: Given the following strings...

// "USA", return true
// "Calvin", return true
// "compUter", return false
// "coding", return true

const capitaliseString = (inputstr) => {
  let allCaps = false,
    noneCap = false,
    firstCap = true;

  for (let i = 0; i < inputstr.length; i++) {
    if (i === 0 && !inputstr[i].toUpperCase() === inputstr[i]) {
      firstCap = false;
      noneCap = true;
    }
    if (i === 1 && inputstr[i].toUpperCase() === inputstr[i]) allCaps = true;
    if (allCaps && inputstr[i].toLowerCase() === inputstr[i]) return false;
    if (noneCap && inputstr[i].toUpperCase() === inputstr[i]) return false;
    if (!i == 0 && !allCaps && inputstr[i].toUpperCase() === inputstr[i])
      return false;
  }
  return true;
};

const test = () => {
  testCases = [
    "USA",
    "Calvin",
    "compUter",
    "coding",
    "CCd",
    "CdC",
    "CCdE",
    "eeC",
    "eCe",
    "eeCe",
  ];
  expecteds = [true, true, false, true, false, false, false, false, false, false];

  for (let i = 0; i < testCases.length; i++) {
    const actual = capitaliseString(testCases[i]);
    console.log(`For '${testCases[i]}' output obtained '${actual}'`);
    console.assert(
      actual === expecteds[i],
      `Wrong answer. Expected: '${expecteds[i]}'. Found: '${actual}'`
    );
  }
  console.log("Test completed.");
};

test();
